### PR TITLE
fix(cheatcodes): rework gas metering pause/resume

### LINKED ISSUE #4523: GasMetering generates different results compared to gasleft() diff
### Component

Forge

### Have you ensured that all of these are up to date?

- [X] Foundry
- [X] Foundryup

### What version of Foundry are you on?

forge 0.2.0 (ed9298d 2023-03-09T00:04:28.809303436Z)

### What command(s) is the bug in?

forge test

### Operating System

Linux

### Describe the bug

In certain scenarios, using `vm.pause/resumeGasMetering()` will yield results that differ from what `gasleft()` would yield.

Below is a simplified reproducible example:
```solidity
pragma solidity ^0.8.18;

import 'forge-std/Test.sol';

contract GasMeterTest is Test {
    mapping(uint => bytes32) map;

    function test_GasMeter () public {
        vm.pauseGasMetering();

        for (uint i = 0; i < 10000; i++) {
            map[i] = keccak256(abi.encode(i));
        }

        vm.resumeGasMetering();

        for (uint i = 0; i < 10000; i++) {
            map[i] = keccak256(abi.encode(i));
        }
    }

    function test_GasLeft () public {
        for (uint i = 0; i < 10000; i++) {
            map[i] = keccak256(abi.encode(i));
        }

        uint start = gasleft();
        for (uint i = 0; i < 10000; i++) {
            map[i] = keccak256(abi.encode(i));
        }
        console2.log("Gas cost:", start - gasleft());
    }
}
```

The output shows the difference:
```
Running 2 tests for test/GasMeterTest.sol:GasMeterTest
[PASS] test_GasLeft() (gas: 231370361)
Logs:
  Gas cost: 6464479

[PASS] test_GasMeter() (gas: 7308777)
Test result: ok. 2 passed; 0 failed; finished in 38.05ms
```

We can see that when calculating the gas used by diffing `gasleft()`, we get `6464479` while using the `vm.pause/resumeGasMetering` cheatcode, we get `7308777`.

I've seen more complex scenarios (that are too long to share) where the difference is considerably larger.

### LINKED ISSUE #5564: `vm.resumeGasMetering` can cause tests with expected reverts to fail with an `OutOfGas` error
### Component

Forge

### Have you ensured that all of these are up to date?

- [X] Foundry
- [X] Foundryup

### What version of Foundry are you on?

forge 0.2.0 (6672134 2023-08-08T00:21:59.824374000Z)

### What command(s) is the bug in?

forge test

### Operating System

macOS (Apple Silicon)

### Describe the bug

After calling `vm.pauseGasMetering`, calling `vm.resumeGasMetering` within a reverting external call causes tests to fail with an `OutOfGas` error. 

There seem to broadly be inconsistencies and unexpected behavior with the `GasMetering` functions; see #5491 as possibly related.

A reproduction, which eliminates self-external-calls as the issue.

```solidity

import {Test, Vm} from "forge-std/Test.sol";

contract RevertingExternalContract {
    error MyError();

    Vm private constant vm = Vm(address(uint160(uint256(keccak256("hevm cheat code")))));

    function externalReverts() external {
        vm.resumeGasMetering();
        revert MyError();
    }
}

contract MeteringTest is Test {
    error MyError();

    RevertingExternalContract externalContract;

    function setUp() public {
        externalContract = new RevertingExternalContract();
    }

    function testSelfNormalRevert() public {
        vm.expectRevert(MeteringTest.MyError.selector);
        this.selfReverts();
    }

    function testSelfMeteringRevert() public {
        vm.pauseGasMetering();
        vm.expectRevert(MeteringTest.MyError.selector);
        this.selfReverts();
    }

    function testExternalNormalRevert() public {
        vm.expectRevert(RevertingExternalContract.MyError.selector);
        externalContract.externalReverts();
    }

    function testExternalMeteringRevert() public {
        vm.pauseGasMetering();
        vm.expectRevert(RevertingExternalContract.MyError.selector);
        externalContract.externalReverts();
    }

    function testExternalMeteringManualCheck() public {
        vm.pauseGasMetering();
        (bool success,) = address(this).call(abi.encodeWithSelector(this.selfReverts.selector));
        assertEq(success, false);
    }

    function selfReverts() external {
        vm.resumeGasMetering();
        revert MyError();
    }
}
```

Output of running these tests:

```
[PASS] testSelfNormalRevert() (gas: 3953)
Test result: FAILED. 2 passed; 3 failed; 0 skipped; finished in 402.42µs
Ran 1 test suites: 2 tests passed, 3 failed, 0 skipped (5 total tests)

Failing tests:
Encountered 3 failing tests in test/Metering.t.sol:MeteringTest
[FAIL. Reason: EvmError: OutOfGas] testExternalMeteringManualCheck() (gas: 9223372036854754743)
[FAIL. Reason: EvmError: OutOfGas] testExternalMeteringRevert() (gas: 9223372036854754743)
[FAIL. Reason: EvmError: OutOfGas] testSelfMeteringRevert() (gas: 9223372036854754743)
```

### LINKED ISSUE #6294: test cannot construct a proxy object while gas metering is on
### Component

Forge

### Have you ensured that all of these are up to date?

- [ ] Foundry
- [ ] Foundryup

### What version of Foundry are you on?

forge 0.2.0 (34f684d 2023-11-01T22:19:24.989266000Z)

### What command(s) is the bug in?

forge test

### Operating System

None

### Describe the bug

The following code reverts in setup.
It does run OK gas metering is NOT disabled
(which means, we can't filter out meter gas only on inner call)
```solidity 
pragma solidity ^0.8.17;

import {Test} from "forge-std/Test.sol";
import "forge-std/Vm.sol";
import {console} from "hardhat/console.sol";
import "@openzeppelin/contracts/access/Ownable.sol";
import "@openzeppelin/contracts/proxy/ERC1967/ERC1967Proxy.sol";
import "@openzeppelin/contracts/proxy/utils/Initializable.sol";
import "@openzeppelin/contracts/proxy/utils/UUPSUpgradeable.sol";

contract MyObject is Ownable, UUPSUpgradeable, Initializable {

    constructor() {
        _disableInitializers();
    }

    function initialize(address anOwner) public virtual initializer {
        _transferOwnership(anOwner);
    }

    function _authorizeUpgrade(address newImplementation) internal override {}
}

contract TestProxy is Test {

    MyObject singleton = new MyObject();

    function setUp() public {
        //with this line enabled, the test revert. comment this line, and the test will pass
        vm.pauseGasMetering();
    }

    function newObj(address owner, uint salt) internal returns (MyObject) {
        return MyObject(address(new ERC1967Proxy{salt: bytes32(salt)}(
            address(singleton),
            abi.encodeCall(MyObject.initialize, (owner))
        )));
    }

    function test_send1() public {
        MyObject obj = newObj(msg.sender, 0);
        vm.resumeGasMetering();
        //this is the method we want to test (and meter for gas)
        obj.owner();
    }
}

```

This is the test result when running with `forge test -vvv`

```sh
[FAIL. Reason: EvmError: Revert] test_send1() (gas: 9223372036854754743)
Traces:
  [0] TestProxy::test_send1()
    ├─ [0] → new <Unknown>@0x3E6e1b3d0069A017c50Cc1d758207728ABb19fdc
    │   ├─ emit Upgraded(implementation: MyObject: [0x5615dEB798BB3E4dFa0139dFa1b3D433Cc23b72f])
    │   ├─ [0] MyObject::initialize(DefaultSender: [0x1804c8AB1F12E6bbf3894d4083f33e07309d1f38]) [delegatecall]
    │   │   ├─ emit OwnershipTransferred(previousOwner: 0x0000000000000000000000000000000000000000, newOwner: DefaultSender: [0x1804c8AB1F12E6bbf3894d4083f33e07309d1f38])
    │   │   ├─ emit Initialized(version: 1)
    │   │   └─ ← ()
    │   └─ ← 0 bytes of code   <<<< this is a constructor call that returns ZERO BYTES
    └─ ← "EvmError: Revert"

```

### PR DESCRIPTION
**Pull request by grandizzy**
## Motivation
closes #5564
closes #6294 - tested with https://github.com/grandizzy/foundry-issue-6294
closes #4523 - optimistically closing, tested with various combinations and wasn't able to repro the big discrepancy anymore (though there's still a difference of ~ 2750 consumed gas by pause / resume calls)

tbd:
#4370 - should be fixed by refunding gas spent by pause / resume calls (similar with diff seen in #4523)

## Solution

- keep track of current execution frame paused gas in `paused_frame_gas` vec  
  - push paused gas when
    - new frame entered (interpreter init) and `pause_gas_metering` flag is true (scenario when gas meter was paused in a prev frame)
    - during interpreter step if `pause_gas_metering` and no gas recorded for current frame (scenario when gas meter was paused in current frame)
  - pop paused gas when any of STOP | RETURN | REVERT | SELFDESTRUCT opcode encountered
- keep gas constant if `pause_gas_metering` is true
- on resume gas metering - set  `pause_gas_metering` to false and remove all recorded paused values

### CODE REVIEW

File: crates/cheatcodes/src/inspector.rs [hunk 1]
```diff
@@ -293,18 +293,14 @@ pub struct Cheatcodes {
...
-    /// paused and creating new contracts.
-    pub gas_metering_create: Option<Option<Gas>>,
+    /// If true then gas metering is paused.
+    pub pause_gas_metering: bool,
+
+    /// Stores paused gas per frame.
+    pub paused_frame_gas: HashMap<usize, Gas>,
+
+    /// Current execution frame, used for pause / resume gas metering.
+    pub current_frame: usize,
```
Comment (DaniPopes) [inline crates/cheatcodes/src/inspector.rs:303]: This looks like it could just be Vec<> to push and pop from.

Comment (grandizzy) [reply crates/cheatcodes/src/inspector.rs]: yep, changed in https://github.com/foundry-rs/foundry/pull/8736/commits/8fb2f5b597ffca603114d2f50d7d726d21a0bb05

**Comment by grandizzy** [comment]
added also fix for clippy

File: crates/forge/tests/cli/test_cmd.rs [hunk 2]
```diff
@@ -1286,3 +1286,83 @@ contract ATest is Test {
...
+
+    function test_GasLeft () public {
+        for (uint i = 0; i < 10000; i++) {
+            map[i] = keccak256(abi.encode(i));
+        }
+
+        uint start = gasleft();
+        for (uint i = 0; i < 10000; i++) {
+            map[i] = keccak256(abi.encode(i));
+        }
```
Comment (DaniPopes) [inline crates/forge/tests/cli/test_cmd.rs:1352]: maybe make this a private function consumeGas and make it take less gas

**Review by DaniPopes — APPROVED** [review]
Much better :+1:

**Pull request merged**